internal/testutil/lsptestutil/lsptestutil.go

```diff
@@ -0,0 +1,339 @@
+package lsptestutil
+
+import (
+	"context"
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/go-json-experiment/json"
+	"github.com/microsoft/typescript-go/internal/bundled"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/ls/lsutil"
+	"github.com/microsoft/typescript-go/internal/lsp"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/project"
+	"github.com/microsoft/typescript-go/internal/project/ata"
+	"github.com/microsoft/typescript-go/internal/project/logging"
+	"github.com/microsoft/typescript-go/internal/testutil/projecttestutil"
+	"github.com/microsoft/typescript-go/internal/vfs"
+	"gotest.tools/v3/assert"
+)
+
+type lspReader struct {
+	c <-chan *lsproto.Message
+}
+
+func (r *lspReader) Read() (*lsproto.Message, error) {
+	msg, ok := <-r.c
+	if !ok {
+		return nil, io.EOF
+	}
+	return msg, nil
+}
+
+type lspWriter struct {
+	c chan<- *lsproto.Message
+}
+
+func (w *lspWriter) Write(msg *lsproto.Message) error {
+	w.c <- msg
+	return nil
+}
+
+func (r *lspWriter) Close() {
+	close(r.c)
+}
+
+var (
+	_ lsp.Reader = (*lspReader)(nil)
+	_ lsp.Writer = (*lspWriter)(nil)
+)
+
+func newLSPPipe() (*lspReader, *lspWriter) {
+	c := make(chan *lsproto.Message, 100)
+	return &lspReader{c: c}, &lspWriter{c: c}
+}
+
+var (
+	ptrTrue                       = ptrTo(true)
+	defaultCompletionCapabilities = &lsproto.CompletionClientCapabilities{
+		CompletionItem: &lsproto.ClientCompletionItemOptions{
+			SnippetSupport:          ptrTrue,
+			CommitCharactersSupport: ptrTrue,
+			PreselectSupport:        ptrTrue,
+			LabelDetailsSupport:     ptrTrue,
+			InsertReplaceSupport:    ptrTrue,
+			DocumentationFormat:     &[]lsproto.MarkupKind{lsproto.MarkupKindMarkdown, lsproto.MarkupKindPlainText},
+		},
+		CompletionList: &lsproto.CompletionListCapabilities{
+			ItemDefaults: &[]string{"commitCharacters", "editRange"},
+		},
+	}
+	defaultDefinitionCapabilities = &lsproto.DefinitionClientCapabilities{
+		LinkSupport: ptrTrue,
+	}
+	defaultTypeDefinitionCapabilities = &lsproto.TypeDefinitionClientCapabilities{
+		LinkSupport: ptrTrue,
+	}
+	defaultHoverCapabilities = &lsproto.HoverClientCapabilities{
+		ContentFormat: &[]lsproto.MarkupKind{lsproto.MarkupKindMarkdown, lsproto.MarkupKindPlainText},
+	}
+)
+
+type TestLspServer struct {
```

Comment:
Nit, but, the Go convention is `TestLSPServer`. Not critical, we already did this elsewhere, so...